### DISCUSSION TITLE
Issues with TLS/SSL

### DISCUSSION THREAD

**Discussion by uhlhosting** [postalserver/postal#1851] [Help with using Postal] [answered]
Hi,

we have the latest version of Postal deployed, since we upgraded to version 2.x and moved from Nginx to Caddy web server, noticed that now the SSLs are not in .pem format for caddy. Thus they cannnot be reused for the TLS feature in SMTP. When I tried to convert existing .key we got this;

```
root@post:/opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/post.uhlhosting.ch# ls -latr
total 28
drwx------ 4 postal users 4096 Nov 24 15:43 ..
-rw------- 1 postal users  227 Jan 29 14:43 post.uhlhosting.ch.key
-rw------- 1 postal users  157 Jan 29 14:43 post.uhlhosting.ch.json
-rw------- 1 postal users 5329 Jan 29 14:43 post.uhlhosting.ch.crt
-rw-r--r-- 1 root   root  1578 Feb  9 12:41 post.uhlhosting.ch.pem
-rw-r--r-- 1 root   root     0 Feb  9 12:44 post.uhlhosting.ch.key.pem
drwx------ 2 postal users 4096 Feb  9 12:44 .
root@post:/opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/post.uhlhosting.ch# rm -rf post.uhlhosting.ch.key.pem 
root@post:/opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/post.uhlhosting.ch# openssl rsa -in post.uhlhosting.ch.key -text > key.pem
139943793616192:error:0607907F:digital envelope routines:EVP_PKEY_get0_RSA:expecting an rsa key:../crypto/evp/p_lib.c:469:
root@post:/opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/post.uhlhosting.ch# 
``

How do you handle Caddy / SMTP ssl in the case when this is not working to convert. Does the config postal.yml only accepts .pem certs and keys? or using .key and .crt will also work?

**Comment by willpower232** [comment]
I don't think it matters what the extension is, you should find they are all text files starting with a certain line of text.

As hinted at by your error message, caddy does not use RSA keys by default so Postal cannot currently handle them. You will either need to change the caddy settings or get a different certificate that is RSA.

**Comment by uhlhosting** [comment]
@willpower232 ca you give us examples of a caddyfile with the settings for a different certificate? i had seen previous posts about this, yet the location where to add the settings, is not mentioned.

**Reply by willpower232** [reply]
I'm afraid I don't know enough about Caddy to comment sorry

**Comment by uhlhosting** [comment]
@willpower232 so we managed with help of Caddy team to add right settings and force caddy to generate RSA certs, I then made them into .pem format, and updated postal.yml, yet it seems when I try SSL/TLS connection it fails.

**Reply by willpower232** [reply]
Have you definitely restarted Postal? Does the key file definitely start with `-----BEGIN RSA PRIVATE KEY-----`?

**Reply by uhlhosting** [reply]
Yes we restarted postal, and they key has the RSA in it.

**Reply by willpower232** [reply]
Sounds like you've done everything correctly. Are you able to watch the SMTP server logs as you make your attempt to see what is going on?

**Reply by uhlhosting** [reply]
I got these weird lines only shown:

```
smtp_1      | /usr/local/bundle/gems/gelf-3.0.0/lib/gelf/notifier.rb:65: warning: constant ::Fixnum is deprecated
smtp_1      | log writing failed. "\xBB" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xBB" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\x9C" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xFF" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xFF" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\x9A" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xC0" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xFF" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\x9A" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xC0" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xFF" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xFC" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\x9C" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xEE" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xC0" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xFF" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\x9A" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xC0" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xFF" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\x9A" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xCD" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xC0" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xFF" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\x92" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xC0" from ASCII-8BIT to UTF-8
smtp_1      | log writing failed. "\xFF" from ASCII-8BIT to UTF-8
root@post:/opt/postal# 
```

**Comment by uhlhosting** [comment]
@willpower232 any hints would help us a great deal.

We had BTCpay server in past working fine with sending mails, now we got:

Isnt there any other ways to debug this?

```
An error occurred while attempting to establish an SSL or TLS connection.

This usually means that the SSL certificate presented by the server is not trusted by the system for one or more of
the following reasons:

1. The server is using a self-signed certificate which cannot be verified.
2. The local system is missing a Root or Intermediate certificate needed to verify the server's certificate.
3. A Certificate Authority CRL server for one or more of the certificates in the chain is temporarily unavailable.
4. The certificate presented by the server is expired or invalid.
5. The set of SSL/TLS protocols supported by the client and server do not match.

See https://github.com/jstedfast/MailKit/blob/master/FAQ.md#SslHandshakeException for possible solutions.
```

We got this in our config:

```
smtp_server:
  port: 25
  tls_enabled: true
  tls_certificate_path: "/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/post.uhlhosting.ch/post.uhlhosting.ch.pem"
  tls_private_key_path: "/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/post.uhlhosting.ch/post.uhlhosting.ch-key.pem"
  ssl_version: TLSv1
  proxy_protocol: false
  log_connect: true
  strip_received_headers: false
  max_message_size: 21

smtp_relays:
  -
    hostname:
    port: 25
    ssl_mode: Auto
```

**Comment by uhlhosting** [comment, marked as answer]
I have changed the path of the certificates in config and it works. Yet would be happy to know why my previous path did not worked.

**Reply by willpower232** [reply]
ah the caddy files aren't in the Postal container unless you put them there manually

**Reply by uhlhosting** [reply]
I think as a workaround will try to use symbolic linking, btw @willpower232 can you tell me if besides beying the certs in RSA, do they need also to be in .pem format? If so then a rule would be need in place, since do caddy generates ssl, yet only RSA not in .pem format. So one would need every 3 months to take the new generated ssl, and or use a own ssl yearly, till its quite annoying process, when we should aim for automation here. If is mandatory to be RSA , maybe could work also in .crt .key formats? Then a symbolic link would suffice from caddy folder to config in postal.

**Reply by willpower232** [reply]
I don't think the extension matters, I've seen them be all sorts. As long as the text content is the right kind, there shouldn't be a problem. I don't know if Caddy would store them differently to how they are generated by certbot.

**Reply by uhlhosting** [reply]
Thanks will give it a try without .pem in this case.

**Reply by onedogsolutions** [reply]
Did symbolic links work? I've tried with the same paths as above and still no TLS.

**Reply by willpower232** [reply]
symbolic links probably won't work since the docker containers filesystem will not see where the original file is so cannot follow the link, duplicating the file would be the only option

**Reply by onedogsolutions** [reply]
Interesting. Still having issues after I copied the certificate and key files.

**Reply by willpower232** [reply]
You might have to use bash inside the container to double check the path to the certificate since I think its a bit different

**Comment by uhlhosting** [comment]
@willpower232 alto I have the certs in RSA format, I just found now a new app that reports me:

`[ERROR] smtp_owner_confirm_hex.go:23 smtpOwnerConfirmHex(): cannot send confirmation email: x509: certificate signed by unknown authority `

**Reply by willpower232** [reply]
The underlying certificate which powers Lets Encrypt changed some months ago, is your new app definitely fully up to date under the hood?

**Reply by uhlhosting** [reply]
@willpower232 
I will ask BTCPay devs, to look into this. Yet not I am into a quest to test all our integraions that were with TLS disabled, to see if it works, it seems I found another one:

```
Unable to connect with TLS encryption Log data: ++ Starting Swift_SmtpTransport << 220 post.uhlhosting.ch ESMTP Postal/FPT4WY >> EHLO ma.uhlhosting.ch << 250-My capabilities are 250-STARTTLS 250 AUTH CRAM-MD5 PLAIN LOGIN >> STARTTLS << 220 Ready to start TLS !! Unable to connect with TLS encryption (code: 0)
++ Starting Swift_SmtpTransport << 220 post.uhlhosting.ch ESMTP Postal/FPT4WY >> EHLO ma.uhlhosting.ch << 250-My capabilities are 250-STARTTLS 250 AUTH CRAM-MD5 PLAIN LOGIN >> STARTTLS << 220 Ready to start TLS !! Unable to connect with TLS encryption (code: 0)
Save changes to the configuration to send a test email.
```

This one is from Mautic. I do run the latest release.

**Reply by onedogsolutions** [reply]
This is the same issue I'm running into. PHPMailer from WordPress. I haven't tested yet, but may try to enable compatible ciphers for my client end during Caddy certificate issue. https://caddy.community/t/tls-no-cipher-suite-supported-by-both-client-and-server-alexa-custom-skill-api-can-not-make-secure-connection-to-caddy-server/11826
